// typia/lib/schemas/metadata/MetadataObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataObject = void 0;
var MetadataObject = (function () {
    function MetadataObject(props) {
        this.nullables = [];
        this.tagged_ = false;
        this.name = props.name;
        this.properties = props.properties;
        this.description = props.description;
        this.jsDocTags = props.jsDocTags;
        this.index = props.index;
        this.validated = props.validated;
        this.recursive = props.recursive;
        this.nullables = [];
        this.tagged_ = false;
    }
    MetadataObject.create = function (props) {
        return new MetadataObject(props);
    };
    MetadataObject._From_without_properties = function (obj) {
        return this.create({
            name: obj.name,
            properties: [],
            description: obj.description,
            jsDocTags: obj.jsDocTags,
            index: obj.index,
            validated: obj.validated,
            recursive: obj.recursive,
            nullables: obj.nullables.slice(),
        });
    };
    MetadataObject.prototype._Messagable = function () {
        return (this.properties.length >= 1 &&
            this.properties.every(function (p) { return p.key.isSoleLiteral(); }));
    };
    MetadataObject.prototype._Is_simple = function (level) {
        if (level === void 0) { level = 0; }
        return (this.recursive === false &&
            this.properties.length < 10 &&
            this.properties.every(function (property) {
                return property.key.isSoleLiteral() &&
                    property.value.size() === 1 &&
                    property.value.isRequired() === true &&
                    property.value.nullable === false &&
                    (property.value.atomics.length === 1 ||
                        (level < 1 &&
                            property.value.objects.length === 1 &&
                            property.value.objects[0]._Is_simple(level + 1)));
            }));
    };
    MetadataObject.prototype.toJSON = function () {
        return {
            name: this.name,
            properties: this.properties.map(function (property) { return property.toJSON(); }),
            description: this.description,
            jsDocTags: this.jsDocTags,
            index: this.index,
            validated: this.validated,
            recursive: this.recursive,
            nullables: this.nullables.slice(),
        };
    };
    return MetadataObject;
}());
exports.MetadataObject = MetadataObject;
(function (MetadataObject) {
    MetadataObject.intersects = function (x, y) {
        return x.properties.some(function (prop) {
            return y.properties.find(function (oppo) { return prop.key.getName() === oppo.key.getName(); }) !== undefined;
        });
    };
    MetadataObject.covers = function (x, y) {
        return x.properties.length >= y.properties.length &&
            x.properties.every(function (prop) {
                return y.properties.find(function (oppo) { return prop.key.getName() === oppo.key.getName(); }) !== undefined;
            });
    };
})(MetadataObject || (exports.MetadataObject = MetadataObject = {}));
