// tgrid/protocols/web/WebServer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebServer = void 0;
var node_1 = require("tstl/utility/node");
var WebAcceptor_1 = require("./WebAcceptor");
var DomainError_1 = require("tstl/exception/DomainError");
var RuntimeError_1 = require("tstl/exception/RuntimeError");
var NodeModule_1 = require("../../utils/internal/NodeModule");
/**
 * Web Socket Server.
 *
 *  - available only in the NodeJS.
 *
 * The `WebServer` is a class who can open an websocket server. Clients connecting to the
 * `WebServer` would communicate with this server through {@link WebAcceptor} objects using
 * RFC (Remote Function Call).
 *
 * To open the websocket server, call the {@link open}() method with your callback function which
 * would be called whenever a {@link WebAcceptor} has been newly created ay a client's connection.
 *
 * Also, when declaring this {@link WebServer} type, you've to define two template arguments,
 * *Header* and *Provider*. The *Header* type repersents an initial data gotten from the remote
 * client after the connection. I hope you and client not to omit it and utilize it as an
 * activation tool to enhance security.
 *
 * The second template argument *Provider* represents the features provided for the remote client.
 * If you don't have any plan to provide any feature to the remote client, just declare it as
 * `null`.
 *
 * @template Header Type of header containing initialization data like activation.
 * @template Provider Type of features provided for the remote systems.
 * @author Jeongho Nam - https://github.com/samchon
 */
var WebServer = /** @class */ (function () {
    function WebServer(key, cert) {
        if ((0, node_1.is_node)() === false)
            throw new DomainError_1.DomainError("Error on WebServer.constructor(): only available in NodeJS.");
        // PREPARE SREVER INSTANCE
        this.options_ = !!key && !!cert ? { key: key, cert: cert } : null;
        // INITIALIZE STATUS & PROTOCOL
        this.state_ = -1 /* WebServer.State.NONE */;
        this.server_ = null;
        this.protocol_ = null;
    }
    /**
     * Open websocket server.
     *
     * Open a server through the web-socket protocol, with its *port* number and *handler*
     * function determining whether to accept the client's connection or not. After the server has
     * been opened, clients can connect to that websocket server by using the {@link WebConnector}
     * class.
     *
     * When implementing the *handler* function with the {@link WebAcceptor} instance, calls the
     * {@link WebAcceptor.accept} method if you want to accept the new client's connection.
     * Otherwise you dont't want to accept the client and reject its connection, just calls the
     * {@link WebAcceptor.reject} instead.
     *
     * @param port Port number to listen.
     * @param handler Callback function for client connection.
     */
    WebServer.prototype.open = function (port, handler) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c;
            var _this = this;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (!(this.state_ === 1 /* WebServer.State.OPEN */)) return [3 /*break*/, 1];
                        throw new DomainError_1.DomainError("Error on WebServer.open(): it has already been opened.");
                    case 1:
                        if (!(this.state_ === 0 /* WebServer.State.OPENING */)) return [3 /*break*/, 2];
                        throw new DomainError_1.DomainError("Error on WebServer.open(): it's on opening, wait for a second.");
                    case 2:
                        if (!(this.state_ === 2 /* WebServer.State.CLOSING */)) return [3 /*break*/, 3];
                        throw new RuntimeError_1.RuntimeError("Error on WebServer.open(): it's on closing.");
                    case 3:
                        if (!(this.server_ === null ||
                            this.state_ === 3 /* WebServer.State.CLOSED */)) return [3 /*break*/, 8];
                        _a = this;
                        if (!(this.options_ !== null)) return [3 /*break*/, 5];
                        return [4 /*yield*/, NodeModule_1.NodeModule.https.get()];
                    case 4:
                        _b = (_d.sent()).createServer(this.options_);
                        return [3 /*break*/, 7];
                    case 5: return [4 /*yield*/, NodeModule_1.NodeModule.http.get()];
                    case 6:
                        _b = (_d.sent()).createServer();
                        _d.label = 7;
                    case 7:
                        _a.server_ = _b;
                        _d.label = 8;
                    case 8:
                        _c = this;
                        return [4 /*yield*/, NodeModule_1.NodeModule.ws.get()];
                    case 9:
                        _c.protocol_ = new (_d.sent()).default.Server({
                            noServer: true,
                        });
                        // SET STATE
                        this.state_ = 0 /* WebServer.State.OPENING */;
                        //----
                        // OPEN SERVER
                        //----
                        // PROTOCOL - ADAPTOR & ACCEPTOR
                        this.server_.on("upgrade", function (request, netSocket, header) {
                            _this.protocol_.handleUpgrade(request, netSocket, header, function (webSocket) {
                                webSocket.once("message", function (data) { return __awaiter(_this, void 0, void 0, function () {
                                    var wrapper, acceptor, exp_1;
                                    return __generator(this, function (_a) {
                                        switch (_a.label) {
                                            case 0:
                                                // @todo: custom code is required
                                                if (typeof data !== "string")
                                                    webSocket.close();
                                                _a.label = 1;
                                            case 1:
                                                _a.trys.push([1, 3, , 4]);
                                                wrapper = JSON.parse(data);
                                                acceptor = WebAcceptor_1.WebAcceptor.create(request, webSocket, wrapper.header);
                                                return [4 /*yield*/, handler(acceptor)];
                                            case 2:
                                                _a.sent();
                                                return [3 /*break*/, 4];
                                            case 3:
                                                exp_1 = _a.sent();
                                                webSocket.close();
                                                return [3 /*break*/, 4];
                                            case 4: return [2 /*return*/];
                                        }
                                    });
                                }); });
                            });
                        });
                        // FINALIZATION
                        return [4 /*yield*/, WebServer._Open(this.server_, port, function (state) { return (_this.state_ = state); })];
                    case 10:
                        // FINALIZATION
                        _d.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Close server.
     *
     * Close all connections between its remote clients ({@link WebConnector}s).
     *
     * It destories all RFCs (remote function calls) between this server and remote clients
     * (through `Driver<Controller>`) that are not returned (completed) yet. The destruction
     * causes all incompleted RFCs to throw exceptions.
     */
    WebServer.prototype.close = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // VALIDATION
                        if (this.state_ !== 1 /* WebServer.State.OPEN */)
                            throw new DomainError_1.DomainError("Error on WebServer.close(): server is not opened.");
                        // DO CLOSE
                        this.state_ = 2 /* WebServer.State.CLOSING */;
                        return [4 /*yield*/, this._Close()];
                    case 1:
                        _a.sent();
                        this.state_ = 3 /* WebServer.State.CLOSED */;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @hidden
     */
    WebServer._Open = function (server, port, setState) {
        return new Promise(function (resolve, reject) {
            // PREPARE RETURNS
            server.on("listening", function () {
                setState(1 /* WebServer.State.OPEN */);
                server.on("error", function () { });
                resolve();
            });
            server.on("error", function (error) {
                setState(-1 /* WebServer.State.NONE */);
                reject(error);
            });
            // DO OPEN - START PROVIDE
            server.listen(port);
        });
    };
    /**
     * @hidden
     */
    WebServer.prototype._Close = function () {
        var _this = this;
        return new Promise(function (resolve) {
            _this.protocol_.close(function () {
                _this.server_.close(function () {
                    resolve();
                });
            });
        });
    };
    Object.defineProperty(WebServer.prototype, "state", {
        /* ----------------------------------------------------------------
            ACCESSORS
        ---------------------------------------------------------------- */
        /**
         * Get server state.
         *
         * Get current state of the websocket server.
         *
         * List of values are such like below:
         *
         *   - `NONE`: The `{@link WebServer} instance is newly created, but did nothing yet.
         *   - `OPENING`: The {@link WebServer.open} method is on running.
         *   - `OPEN`: The websocket server is online.
         *   - `CLOSING`: The {@link WebServer.close} method is on running.
         *   - `CLOSED`: The websocket server is offline.
         */
        get: function () {
            return this.state_;
        },
        enumerable: false,
        configurable: true
    });
    return WebServer;
}());
exports.WebServer = WebServer;
/**
 *
 */
(function (WebServer) {
})(WebServer || (exports.WebServer = WebServer = {}));
