// typia/lib/functional/$ProtobufReader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.$ProtobufReader = void 0;
var $ProtobufReader = (function () {
    function $ProtobufReader(buf) {
        this.buf = buf;
        this.ptr = 0;
        this.view = new DataView(buf.buffer);
    }
    $ProtobufReader.prototype.index = function () {
        return this.ptr;
    };
    $ProtobufReader.prototype.size = function () {
        return this.buf.length;
    };
    $ProtobufReader.prototype.uint32 = function () {
        return this.varint32();
    };
    $ProtobufReader.prototype.int32 = function () {
        return this.varint32();
    };
    $ProtobufReader.prototype.sint32 = function () {
        var value = this.varint32();
        return (value >>> 1) ^ -(value & 1);
    };
    $ProtobufReader.prototype.uint64 = function () {
        return this.varint64();
    };
    $ProtobufReader.prototype.int64 = function () {
        return this.varint64();
    };
    $ProtobufReader.prototype.sint64 = function () {
        var value = this.varint64();
        return (value >> N01) ^ -(value & N01);
    };
    $ProtobufReader.prototype.bool = function () {
        return this.varint32() !== 0;
    };
    $ProtobufReader.prototype.float = function () {
        var value = this.view.getFloat32(this.ptr, true);
        this.ptr += 4;
        return value;
    };
    $ProtobufReader.prototype.double = function () {
        var value = this.view.getFloat64(this.ptr, true);
        this.ptr += 8;
        return value;
    };
    $ProtobufReader.prototype.bytes = function () {
        var length = this.uint32();
        var from = this.ptr;
        this.ptr += length;
        return this.buf.subarray(from, from + length);
    };
    $ProtobufReader.prototype.string = function () {
        return utf8.decode(this.bytes());
    };
    $ProtobufReader.prototype.skip = function (length) {
        if (length === 0)
            while (this.u8() & 0x80)
                ;
        else {
            if (this.index() + length > this.size())
                throw new Error("Error on typia.protobuf.decode(): buffer overflow.");
            this.ptr += length;
        }
    };
    $ProtobufReader.prototype.skipType = function (wireType) {
        switch (wireType) {
            case 0:
                this.skip(0);
                break;
            case 1:
                this.skip(8);
                break;
            case 2:
                this.skip(this.uint32());
                break;
            case 3:
                while ((wireType = this.uint32() & 0x07) !== 4)
                    this.skipType(wireType);
                break;
            case 5:
                this.skip(4);
                break;
            default:
                throw new Error("Invalid wire type ".concat(wireType, " at offset ").concat(this.ptr, "."));
        }
    };
    $ProtobufReader.prototype.varint32 = function () {
        var loaded;
        var value;
        value = (loaded = this.u8()) & 0x7f;
        if (loaded < 0x80)
            return value;
        value |= ((loaded = this.u8()) & 0x7f) << 7;
        if (loaded < 0x80)
            return value;
        value |= ((loaded = this.u8()) & 0x7f) << 14;
        if (loaded < 0x80)
            return value;
        value |= ((loaded = this.u8()) & 0x7f) << 21;
        if (loaded < 0x80)
            return value;
        value |= ((loaded = this.u8()) & 0xf) << 28;
        if (loaded < 0x80)
            return value;
        if (this.u8() < 0x80)
            return value;
        if (this.u8() < 0x80)
            return value;
        if (this.u8() < 0x80)
            return value;
        if (this.u8() < 0x80)
            return value;
        if (this.u8() < 0x80)
            return value;
        return value;
    };
    $ProtobufReader.prototype.varint64 = function () {
        var loaded;
        var value;
        value = (loaded = this.u8n()) & N7F;
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(7);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(14);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(21);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(28);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(35);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(42);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(49);
        if (loaded < N80)
            return value;
        value |= ((loaded = this.u8n()) & N7F) << BigInt(56);
        if (loaded < N80)
            return value;
        value |= (this.u8n() & N01) << BigInt(63);
        return BigInt.asIntN(64, value);
    };
    $ProtobufReader.prototype.u8 = function () {
        return this.view.getUint8(this.ptr++);
    };
    $ProtobufReader.prototype.u8n = function () {
        return BigInt(this.u8());
    };
    return $ProtobufReader;
}());
exports.$ProtobufReader = $ProtobufReader;
var utf8 = new TextDecoder();
var N01 = BigInt(0x01);
var N7F = BigInt(0x7f);
var N80 = BigInt(0x80);
