// typia/lib/functional/$ProtobufSizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.$ProtobufSizer = void 0;
var _strlen_1 = require("./$strlen");
var $ProtobufSizer = (function () {
    function $ProtobufSizer(length) {
        if (length === void 0) { length = 0; }
        this.len = length;
        this.pos = [];
        this.varlen = [];
        this.varlenidx = [];
    }
    $ProtobufSizer.prototype.bool = function () {
        this.len += 1;
    };
    $ProtobufSizer.prototype.int32 = function (value) {
        if (value < 0) {
            this.len += 10;
        }
        else {
            this.varint32(value);
        }
    };
    $ProtobufSizer.prototype.sint32 = function (value) {
        this.varint32((value << 1) ^ (value >> 31));
    };
    $ProtobufSizer.prototype.uint32 = function (value) {
        this.varint32(value);
    };
    $ProtobufSizer.prototype.int64 = function (value) {
        this.varint64(typeof value === "number" ? BigInt(value) : value);
    };
    $ProtobufSizer.prototype.sint64 = function (value) {
        if (typeof value === "number")
            value = BigInt(value);
        this.varint64((value << BigInt(1)) ^ (value >> BigInt(63)));
    };
    $ProtobufSizer.prototype.uint64 = function (value) {
        this.varint64(typeof value === "number" ? BigInt(value) : value);
    };
    $ProtobufSizer.prototype.float = function (_value) {
        this.len += 4;
    };
    $ProtobufSizer.prototype.double = function (_value) {
        this.len += 8;
    };
    $ProtobufSizer.prototype.bytes = function (value) {
        this.uint32(value.byteLength);
        this.len += value.byteLength;
    };
    $ProtobufSizer.prototype.string = function (value) {
        var len = (0, _strlen_1.$strlen)(value);
        this.varlen.push(len);
        this.uint32(len);
        this.len += len;
    };
    $ProtobufSizer.prototype.fork = function () {
        this.pos.push(this.len);
        this.varlenidx.push(this.varlen.length);
        this.varlen.push(0);
    };
    $ProtobufSizer.prototype.ldelim = function () {
        if (!(this.pos.length && this.varlenidx.length))
            throw new Error("Error on typia.protobuf.encode(): missing fork() before ldelim() call.");
        var endPos = this.len;
        var startPos = this.pos.pop();
        var idx = this.varlenidx.pop();
        var len = endPos - startPos;
        this.varlen[idx] = len;
        this.uint32(len);
    };
    $ProtobufSizer.prototype.reset = function () {
        this.len = 0;
        this.pos.length = 0;
        this.varlen.length = 0;
        this.varlenidx.length = 0;
    };
    $ProtobufSizer.prototype.varint32 = function (value) {
        this.len +=
            value < 0
                ? 10
                : value < 0x80
                    ? 1
                    : value < 0x4000
                        ? 2
                        : value < 0x200000
                            ? 3
                            : value < 0x10000000
                                ? 4
                                : 5;
    };
    $ProtobufSizer.prototype.varint64 = function (val) {
        val = BigInt.asUintN(64, val);
        while (val > NX7F) {
            ++this.len;
            val = val >> ND07;
        }
        ++this.len;
    };
    return $ProtobufSizer;
}());
exports.$ProtobufSizer = $ProtobufSizer;
var ND07 = BigInt(7);
var NX7F = BigInt(0x7f);
