// @nestia/core/lib/decorators/internal/validate_request_body.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate_request_body = void 0;
var common_1 = require("@nestjs/common");
var typia_1 = __importDefault(require("typia"));
var NoTransformConfigureError_1 = require("./NoTransformConfigureError");
var validate_request_body = function (method) {
    return function (validator) {
        if (!validator)
            return function () { return (0, NoTransformConfigureError_1.NoTransformConfigureError)(method); };
        else if (validator.type === "assert")
            return assert(validator.assert);
        else if (validator.type === "is")
            return is(validator.is);
        else if (validator.type === "validate")
            return validate(validator.validate);
        return function () {
            return new Error("Error on nestia.core.".concat(method, "(): invalid typed validator."));
        };
    };
};
exports.validate_request_body = validate_request_body;
var assert = function (closure) {
    return function (input) {
        try {
            closure(input);
            return null;
        }
        catch (exp) {
            if ((function (input) {
                var $io0 = function (input) { return "string" === typeof input.method && (undefined === input.path || "string" === typeof input.path) && "string" === typeof input.expected && true && "string" === typeof input.name && "string" === typeof input.message && (undefined === input.stack || "string" === typeof input.stack); };
                return "object" === typeof input && null !== input && $io0(input);
            })(exp)) {
                return new common_1.BadRequestException({
                    path: exp.path,
                    reason: exp.message,
                    expected: exp.expected,
                    value: exp.value,
                    message: MESSAGE,
                });
            }
            throw exp;
        }
    };
};
var is = function (closure) {
    return function (input) {
        var success = closure(input);
        return success ? null : new common_1.BadRequestException(MESSAGE);
    };
};
var validate = function (closure) {
    return function (input) {
        var result = closure(input);
        return result.success
            ? null
            : new common_1.BadRequestException({
                errors: result.errors,
                message: MESSAGE,
            });
    };
};
var MESSAGE = "Request body data is not following the promised type.";
