// typia/lib/factories/NumericRangeFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumericRangeFactory = void 0;
var ExpressionFactory_1 = require("./ExpressionFactory");
var NumericRangeFactory;
(function (NumericRangeFactory) {
    NumericRangeFactory.number = function (context) {
        return function (type) {
            return function (input) {
                return ExpressionFactory_1.ExpressionFactory.transpile(context)(NUMBER_TEXT[type])(input);
            };
        };
    };
    NumericRangeFactory.bigint = function (context) {
        return function (type) {
            return function (input) {
                return ExpressionFactory_1.ExpressionFactory.transpile(context)(BIGINT_TEXT[type])(input);
            };
        };
    };
})(NumericRangeFactory || (exports.NumericRangeFactory = NumericRangeFactory = {}));
var NUMBER_TEXT = {
    int32: "Math.floor($input) === $input && -2147483648 <= $input && $input <= 2147483647",
    uint32: "Math.floor($input) === $input && 0 <= $input && $input <= 4294967295",
    int64: "Math.floor($input) === $input && -9223372036854775808 <= $input && $input <= 9223372036854775807",
    uint64: "Math.floor($input) === $input && 0 <= $input && $input <= 18446744073709551615",
    float: "-1.175494351e38 <= $input && $input <= 3.4028235e38",
    double: "true",
};
var BIGINT_TEXT = {
    int64: "true",
    uint64: "BigInt(0) <= $input",
};
