// typia/lib/protobuf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createValidateEncode = exports.createAssertEncode = exports.createIsEncode = exports.createEncode = exports.createValidateDecode = exports.createAssertDecode = exports.createIsDecode = exports.createDecode = exports.validateEncode = exports.isEncode = exports.assertEncode = exports.encode = exports.validateDecode = exports.isDecode = exports.assertDecode = exports.decode = exports.message = void 0;
var Namespace_1 = require("./functional/Namespace");
function message() {
    halt("message");
}
exports.message = message;
function decode() {
    halt("decode");
}
exports.decode = decode;
Object.assign(decode, Namespace_1.Namespace.protobuf.decode("decode"));
function assertDecode() {
    halt("assertDecode");
}
exports.assertDecode = assertDecode;
Object.assign(assertDecode, Namespace_1.Namespace.assert("protobuf.assertDecode"));
Object.assign(assertDecode, Namespace_1.Namespace.protobuf.decode("assertDecode"));
function isDecode() {
    halt("isDecode");
}
exports.isDecode = isDecode;
Object.assign(isDecode, Namespace_1.Namespace.is());
Object.assign(isDecode, Namespace_1.Namespace.protobuf.decode("isDecode"));
function validateDecode() {
    halt("validateDecode");
}
exports.validateDecode = validateDecode;
Object.assign(validateDecode, Namespace_1.Namespace.validate());
Object.assign(validateDecode, Namespace_1.Namespace.protobuf.decode("validateDecode"));
function encode() {
    halt("encode");
}
exports.encode = encode;
Object.assign(encode, Namespace_1.Namespace.protobuf.encode("encode"));
function assertEncode() {
    halt("assertEncode");
}
exports.assertEncode = assertEncode;
Object.assign(assertEncode, Namespace_1.Namespace.assert("protobuf.assertEncode"));
Object.assign(assertEncode, Namespace_1.Namespace.protobuf.encode("assertEncode"));
function isEncode() {
    halt("isEncode");
}
exports.isEncode = isEncode;
Object.assign(isEncode, Namespace_1.Namespace.is());
Object.assign(isEncode, Namespace_1.Namespace.protobuf.encode("isEncode"));
function validateEncode() {
    halt("validateEncode");
}
exports.validateEncode = validateEncode;
Object.assign(validateEncode, Namespace_1.Namespace.validate());
Object.assign(validateEncode, Namespace_1.Namespace.protobuf.encode("validateEncode"));
function createDecode() {
    halt("createDecode");
}
exports.createDecode = createDecode;
Object.assign(createDecode, Namespace_1.Namespace.protobuf.decode("createDecode"));
function createIsDecode() {
    halt("createIsDecode");
}
exports.createIsDecode = createIsDecode;
Object.assign(createIsDecode, Namespace_1.Namespace.is());
Object.assign(createIsDecode, Namespace_1.Namespace.protobuf.decode("createIsDecode"));
function createAssertDecode() {
    halt("createAssertDecode");
}
exports.createAssertDecode = createAssertDecode;
Object.assign(createAssertDecode, Namespace_1.Namespace.assert("protobuf.createAssertDecode"));
Object.assign(createAssertDecode, Namespace_1.Namespace.protobuf.decode("createAssertDecode"));
function createValidateDecode() {
    halt("createValidateDecode");
}
exports.createValidateDecode = createValidateDecode;
Object.assign(createValidateDecode, Namespace_1.Namespace.validate());
Object.assign(createValidateDecode, Namespace_1.Namespace.protobuf.decode("createValidateDecode"));
function createEncode() {
    halt("createEncode");
}
exports.createEncode = createEncode;
Object.assign(createEncode, Namespace_1.Namespace.protobuf.encode("createEncode"));
function createIsEncode() {
    halt("createIsEncode");
}
exports.createIsEncode = createIsEncode;
Object.assign(createIsEncode, Namespace_1.Namespace.is());
Object.assign(createIsEncode, Namespace_1.Namespace.protobuf.encode("createIsEncode"));
function createAssertEncode() {
    halt("createAssertEncode");
}
exports.createAssertEncode = createAssertEncode;
Object.assign(createAssertEncode, Namespace_1.Namespace.assert("protobuf.createAssertEncode"));
Object.assign(createAssertEncode, Namespace_1.Namespace.protobuf.encode("createAssertEncode"));
function createValidateEncode() {
    halt("createValidateEncode");
}
exports.createValidateEncode = createValidateEncode;
Object.assign(createValidateEncode, Namespace_1.Namespace.validate());
Object.assign(createValidateEncode, Namespace_1.Namespace.protobuf.encode("createValidateEncode"));
function halt(name) {
    throw new Error("Error on typia.protobuf.".concat(name, "(): no transform has been configured. Read and follow https://typia.io/docs/setup please."));
}
