// tgrid/protocols/workers/internal/FileSystem.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystem = void 0;
var NodeModule_1 = require("../../../utils/internal/NodeModule");
/**
 * @hidden
 */
var FileSystem;
(function (FileSystem) {
    /* ----------------------------------------------------------------
        ACCESSORS
    ---------------------------------------------------------------- */
    function exists(path) {
        return __awaiter(this, void 0, void 0, function () {
            var exists;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        exists = (_a.sent()).exists;
                        return [2 /*return*/, new Promise(function (resolve) {
                                exists(path, resolve);
                            })];
                }
            });
        });
    }
    FileSystem.exists = exists;
    function dir(path) {
        return __awaiter(this, void 0, void 0, function () {
            var readdir;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        readdir = (_a.sent()).readdir;
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                readdir(path, function (err, ret) {
                                    if (err)
                                        reject(err);
                                    else
                                        resolve(ret);
                                });
                            })];
                }
            });
        });
    }
    FileSystem.dir = dir;
    function lstat(path) {
        return __awaiter(this, void 0, void 0, function () {
            var lstat;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        lstat = (_a.sent()).lstat;
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                lstat(path, function (err, stat) {
                                    if (err)
                                        reject(err);
                                    else
                                        resolve(stat);
                                });
                            })];
                }
            });
        });
    }
    FileSystem.lstat = lstat;
    function read(path, encoding) {
        return __awaiter(this, void 0, void 0, function () {
            var readFile;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        readFile = (_a.sent()).readFile;
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                var callback = function (err, ret) {
                                    if (err)
                                        reject(err);
                                    else
                                        resolve(ret);
                                };
                                if (encoding === undefined)
                                    readFile(path, callback);
                                else
                                    readFile(path, encoding, callback);
                            })];
                }
            });
        });
    }
    FileSystem.read = read;
    /* ----------------------------------------------------------------
        ARCHIVERS
    ---------------------------------------------------------------- */
    function mkdir(path) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, exists(path)];
                    case 1:
                        if (!((_a.sent()) === false)) return [3 /*break*/, 3];
                        return [4 /*yield*/, _Mkdir(path)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    }
    FileSystem.mkdir = mkdir;
    function _Mkdir(path) {
        return __awaiter(this, void 0, void 0, function () {
            var mkdir;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        mkdir = (_a.sent()).mkdir;
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                mkdir(path, function (err) {
                                    if (err)
                                        reject(err);
                                    else
                                        resolve();
                                });
                            })];
                }
            });
        });
    }
    function write(path, content) {
        return __awaiter(this, void 0, void 0, function () {
            var writeFile;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        writeFile = (_a.sent()).writeFile;
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                var callback = function (err) {
                                    if (err)
                                        reject(err);
                                    else
                                        resolve();
                                };
                                if (content instanceof Buffer)
                                    writeFile(path, content, callback);
                                else
                                    writeFile(path, content, "utf8", callback);
                            })];
                }
            });
        });
    }
    FileSystem.write = write;
    function unlink(path) {
        return __awaiter(this, void 0, void 0, function () {
            var unlink;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, NodeModule_1.NodeModule.fs.get()];
                    case 1:
                        unlink = (_a.sent()).unlink;
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                unlink(path, function (err) {
                                    if (err)
                                        reject(err);
                                    else
                                        resolve();
                                });
                            })];
                }
            });
        });
    }
    FileSystem.unlink = unlink;
})(FileSystem || (exports.FileSystem = FileSystem = {}));
