// nestjs-pino/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PARAMS_PROVIDER_TOKEN = exports.LoggerErrorInterceptor = exports.getLoggerToken = exports.InjectPinoLogger = exports.PinoLogger = exports.Logger = exports.LoggerModule = void 0;
var LoggerModule_1 = require("./LoggerModule");
Object.defineProperty(exports, "LoggerModule", { enumerable: true, get: function () { return LoggerModule_1.LoggerModule; } });
var Logger_1 = require("./Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return Logger_1.Logger; } });
var PinoLogger_1 = require("./PinoLogger");
Object.defineProperty(exports, "PinoLogger", { enumerable: true, get: function () { return PinoLogger_1.PinoLogger; } });
var InjectPinoLogger_1 = require("./InjectPinoLogger");
Object.defineProperty(exports, "InjectPinoLogger", { enumerable: true, get: function () { return InjectPinoLogger_1.InjectPinoLogger; } });
Object.defineProperty(exports, "getLoggerToken", { enumerable: true, get: function () { return InjectPinoLogger_1.getLoggerToken; } });
var LoggerErrorInterceptor_1 = require("./LoggerErrorInterceptor");
Object.defineProperty(exports, "LoggerErrorInterceptor", { enumerable: true, get: function () { return LoggerErrorInterceptor_1.LoggerErrorInterceptor; } });
var params_1 = require("./params");
Object.defineProperty(exports, "PARAMS_PROVIDER_TOKEN", { enumerable: true, get: function () { return params_1.PARAMS_PROVIDER_TOKEN; } });
